src/libpriv/rpmostree-core.cxx

```diff
@@ -679,6 +682,28 @@ core_libdnf_process_global_init()
 
 } /* namespace */
 
+// Union the set of packages requested via treespec and via treefile.
+// TODO remove this once treespec is gone, then we just can iterate over `self->treefile_rs->get_packages()`
+static rust::Vec<rust::String>
+get_requested_packages (RpmOstreeContext *self)
+{
+  rust::Vec<rust::String> packages;
+  { g_autofree char **pkgnames_c = NULL;
+    g_variant_dict_lookup (self->spec->dict, "packages", "^a&s", &pkgnames_c);
+    for (char **it = pkgnames_c; it && *it; it++)
+      packages.push_back(*it);
+  }
+  if (self->treefile_rs)
+    {
+      for (auto &pkgname : self->treefile_rs->get_packages())
+        {
+          packages.push_back(pkgname);
```

Comment:
Are duplicate packages (i.e. same entry coming from both treespec and treefile) a concern here? Should this vector be sorted and deduplicated at this construction site myabe?

Reply:
Hmm.  Duplicate packages wouldn't break anything, it's like listing a package twice in a `manifest.yaml`.   *But* now that you mention it, I think actually treefile and treespec should be mutually exclusive, which would simplify the code.

I think we're almost in that state, except for `rpm-ostree compose extensions`.  I need to look at that.